### PR TITLE
Fixes #2665 - WireMock.stubFor throws descriptive exception message on 404 error code from non-WireMock server instead of unclear JsonException

### LINKED ISSUE #2665: WireMock.stubFor(..) should have  appropriate error description instead of unclear JsonException on 404 error code from non-WireMock server
### Proposal

I have Spring-Boot application and integration tests using Cucumber and WireMock. When switching tests to parallel  execution mode, I started to receive the following unclear exceptions during invocation of `WireMock.stubFor(..)`:

```
com.github.tomakehurst.wiremock.common.JsonException: 
{
  "errors" : [ {
    "code" : 10,
    "source" : {
      "pointer" : "/code"
    },
    "title" : "Error parsing JSON",
    "detail" : "Unrecognized field \"code\" (class com.github.tomakehurst.wiremock.common.Errors), not marked as ignorable"
  } ]
}
    at com.github.tomakehurst.wiremock.common.JsonException.fromJackson(JsonException.java:53)
    at com.github.tomakehurst.wiremock.common.Json.read(Json.java:60)
    at com.github.tomakehurst.wiremock.client.HttpAdminClient.safelyExecuteRequest(HttpAdminClient.java:519)
    at com.github.tomakehurst.wiremock.client.HttpAdminClient.executeRequest(HttpAdminClient.java:489)
    at com.github.tomakehurst.wiremock.client.HttpAdminClient.executeRequest(HttpAdminClient.java:466)
    at com.github.tomakehurst.wiremock.client.HttpAdminClient.addStubMapping(HttpAdminClient.java:146)
    at com.github.tomakehurst.wiremock.client.WireMock.register(WireMock.java:414)
    at com.github.tomakehurst.wiremock.client.WireMock.register(WireMock.java:409)
    at com.github.tomakehurst.wiremock.client.WireMock.givenThat(WireMock.java:115)
    at com.github.tomakehurst.wiremock.client.WireMock.stubFor(WireMock.java:119)
```

The stack trace is from version: `com.github.tomakehurst:wiremock-jre8-standalone:2.33.0`, but code logic inside `HttpAdminClient` with the latest version is the same.

WireMock was created from thread `main` in the following way:
```
WireMockServer wireMock = new WireMockServer(options()
    .port(8000)
    .extensions(new ResponseTemplateTransformer(true))
    .notifier(new ConsoleNotifier(true)));
wireMock.start();
configureFor(8000);
```

Stub initialization:
```
WireMock.stubFor(post(urlPathMatching("/test"))
    .withRequestBody(equalToJson(..))
    .willReturn(aResponse()
        .withStatus(200)
        .withHeader("Content-Type", "application/json")
        .withBody(..)));
```

After investigation,  it was discovered that the response body during stub definition execution is `{"code":"404.null.55","message":"Not Found"}`. The issue was that request targeted the wrong port, it took a default port 8080 instead of configured 8000. It's due to `WireMock.stubFor(..)` is invoked from another thread (due to parallel execution mode by Cucumber).  Due to implementation inside `WireMock`, it has thread local `defaultInstance`, which should be additionally configured with appropriate port from the invoked thread.
Port 8000 was configured for `WireMock` server and 8080 for Tomcat from Spring Boot test.

If there is no process on port 8080, it will throw `HttpHostConnectException` with url (including port number) as part of exception message, so it helps investigate various issues.

So I updated code to invoke `WireMock.configureFor(8000);` before each invocation of `WireMock.stubFor(..)` to guarantee that port is configured per each required thread, and it fixed the issue.
The initial exception that I faced as `JsonException Unrecognized field 'code'` is due to not properly handling 404 client error code inside [HttpAdminClient](https://github.com/wiremock/wiremock/blob/2.33.0/src/main/java/com/github/tomakehurst/wiremock/client/HttpAdminClient.java#L518-L519), as this 404 error not from Wire Mock server, but might be from any other server (indeed it's a rare case, but still might happen). 

It would be great to have an appropriate error description (with an explicitly specified url that includes port) instead of unclear `JsonException` for this use case.

### Reproduction steps

Start Spring-Boot application on port 8080.
Then configure WireMock server without invoking `configureFor(..);`:
```
WireMockServer wireMock = new WireMockServer(options()
    .port(8000)
    .extensions(new ResponseTemplateTransformer(true))
    .notifier(new ConsoleNotifier(true)));
wireMock.start();
```

When `stubFor(..)` is invoked:
```
WireMock.stubFor(post(urlPathMatching("/test"))
    .withRequestBody(equalToJson(..))
    .willReturn(aResponse()
        .withStatus(200)
        .withHeader("Content-Type", "application/json")
        .withBody(..)));
```

non-descriptive exception `JsonException` will be thrown

### References

I will provide proposed pull request

### PR DESCRIPTION
**Pull request by vasiliy-sarzhynskyi**
WireMock.stubFor(..) should throw descriptive exception message on 404 error code from non-WireMock server instead of an unclear JsonException

## References

- https://github.com/wiremock/wiremock/issues/2665

## Submitter checklist

- [x] Recommended: Join [WireMock Slack](https://slack.wiremock.org/) to get any help in `#help-contributing` or a project-specific channel like `#wiremock-java`
- [x] The PR request is well described and justified, including the body and the references
- [x] The PR title represents the desired changelog entry
- [x] The repository's code style is followed (see the contributing guide)
- [x] Test coverage that demonstrates that the change works as expected
- [x] For new features, there's necessary documentation in this pull request or in a subsequent PR to [wiremock.org](https://github.com/wiremock/wiremock.org)

### CODE REVIEW

**Comment by dieppa** [comment]
Current output:


```
com.github.tomakehurst.wiremock.common.JsonException: {
  "errors" : [ {
    "code" : 10,
    "source" : { },
    "title" : "Error parsing JSON",
    "detail" : "Unexpected character ('<' (code 60)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')\n at [Source: (String)\"<h1>404 Not Found</h1>No context found for request\"; line: 1, column: 1]"
  } ]
}
```

File: src/test/java/com/github/tomakehurst/wiremock/client/HttpAdminClientTest.java [hunk 1]
```diff
@@ -91,4 +95,17 @@ public void reuseConnections() throws InterruptedException, IOException {
...
+      throws IOException {
+    var nonWireMockServer = HttpServer.create(new InetSocketAddress(0), 0);
+    nonWireMockServer.start();
+    var serverPort = nonWireMockServer.getAddress().getPort();
+    var client = new HttpAdminClient("localhost", serverPort, ADMIN_TEST_PREFIX);
+    var mapping = post(urlPathMatching("/test")).willReturn(ok()).build();
+    var thrown = assertThrows(ClientError.class, () -> client.addStubMapping(mapping));
+    assertThat(thrown.getMessage()).contains("localhost:" + serverPort);
+    nonWireMockServer.stop(0);
+  }
```
Comment (dieppa) [inline src/test/java/com/github/tomakehurst/wiremock/client/HttpAdminClientTest.java:110]: What about adjusting the test to check the error message does actually match the expectation, further than just checking that contains `localhost`?

**Review by dieppa — CHANGES_REQUESTED** [review]
Hello @vasiliy-sarzhynskyi , your change looks good and it seems that provides value without adding a breaking change; it doesn't change the error structure, just the error message, which is ok.

However, this is a minor change I would like to have in the tests before merging it.

Thanks in advance!

File: src/test/java/com/github/tomakehurst/wiremock/client/HttpAdminClientTest.java (hunk 1 above)
Comment (vasiliy-sarzhynskyi) [reply src/test/java/com/github/tomakehurst/wiremock/client/HttpAdminClientTest.java]: I extended assertions for thrown exception. I don't want to compare the whole message, cause part of it depends on non WireMock server and is not controlled by this repo, so checked exception details by regex.

e.g. creating a server by `HttpServer.create(..)`, we get the following message:

`Error parsing response body '<h1>404 Not Found</h1>No context found for request' with status code 404 for http://localhost:xyz/admin-test/__admin/mappings. Error: Unexpected character ('<' (code 60)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')\n at [Source: (String)\"<h1>404 Not Found</h1>No context found for request\"; line: 1, column: 1]`


but with Spring-Boot server in will be absolutely different one:

`Error parsing response body '{\"code\":\"404.null.55\",\"message\":\"Not Found\"}' with status code 404 for http://localhost:xyz/admin-test/__admin/mappings. Error: Unrecognized field \"code\", not marked as ignorable.
`

**Review by dieppa — APPROVED** [review]

**Pull request merged**